streaming/stream_transfer_task.cc

```diff
@@ -199,18 +199,27 @@ future<> send_mutation_fragments(lw_shared_ptr<send_info> si) {
         }();
 
         auto sink_op = [sink, si, got_error_from_peer] () mutable -> future<> {
-            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment>& sink) {
+            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink) {
                 return repeat([&sink, si, got_error_from_peer] () mutable {
                     return si->reader(db::no_timeout).then([&sink, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
                         if (mf && !(*got_error_from_peer)) {
                             frozen_mutation_fragment fmf = freeze(*s, *mf);
                             auto size = fmf.representation().size();
                             streaming::get_local_stream_manager().update_progress(si->plan_id, si->id.addr, streaming::progress_info::direction::OUT, size);
-                            return sink(fmf).then([] { return stop_iteration::no; });
+                            return sink(fmf, stream_mutation_fragments_cmd::mutation_fragment_data).then([] { return stop_iteration::no; });
                         } else {
                             return make_ready_future<stop_iteration>(stop_iteration::yes);
                         }
                     });
+                }).then([&sink] () mutable {
+                    return sink(frozen_mutation_fragment(bytes_ostream()), stream_mutation_fragments_cmd::end_of_stream);
+                }).then([&sink] () mutable {
+                    return sink.flush();
```

Comment:
Why do we need to `flush()`?

Reply:
Make sure the sender does not cache and flush the sending queue because there is nothing to send. It is currently a no-op but it might change in the future.

Reply:
I still don't understand why we need it. What bad would happen without it?

Reply:
Data may be buffered indefinitely, so if sending new data depends on a reply from a receiver and the receiver replies only when all data is received (ping-pong pattern) there will be deadlock.

Reply:
@gleb-cloudius We don't wait for anything after `flush()`here, we call `close()` right after calling `flush()`. Is the flush still needed in such a case?

Reply:
No need to flush before close, only if we are going to reuse the connection to stream more data later.

Reply:
So @asias, please remove the flush() as a follow up

Reply:
I wanted to make it future-proof if we use reuse the connection inside the rpc framework in the future. 
But since we call close(), even if we reuse the connection, the rpc should flush the data before reusing the connection.

I will remove the flush().

It is better to squash the change in this commit because we want to backport to 3.0 and 3.0 does not have flush in seastar. 

But it is too late, I saw it is merged in master.

So, let's commit the patch to remove flush and backport two of the commit in a single commit to 3.0.